### PR TITLE
ocean depth fix for cnrm_esm2_1, gfdl_esm4, ipsl_cm6a_lr datasets +  mcm_ua_1_0

### LINKED ISSUE #1069: Apply ocean depth fixes for cnrm_esm2_1, gfdl_esm4, ipsl_cm6a_lr
This activity relates to the development of ESMValTool recipes from CE2COAST project (https://github.com/ESMValGroup/ESMValTool/issues/2114) that require a number of fixes in ESMValCore.
The recipe(s) from CE2COAST project deal primarily with **oceanic parameters**, e.g., thetao, no3, po4 ...

1. There is the need to **propagate the time coordinate also after the climate statistics** to correctly create a sequence of preprocessing operations (see details in issue https://github.com/ESMValGroup/ESMValCore/issues/1018)

2. A number of datasets with incorrect definition of ocean depth coordinate in `Omon` files:
- **cnrm_esm2_1, gfdl_esm4, ipsl_cm6a_lr** : have wrong standard name of level coordinate
- **cesm2**: have wrong standard name of level coordinate and depth coordinate is in `centimeters` instead of `meters`, but not in level bounds (see also https://github.com/ESMValGroup/ESMValCore/issues/859 and https://github.com/ESMValGroup/ESMValCore/pull/888)

As the issue with depth coordinate affect most of 3D variables in `Omon`, it would be ideal to create a fix that operate at the mip level, instead of declaring a fix for each variable. 
@valeriupredoi do you think this could be a viable solution?

### PR DESCRIPTION
**Pull request by tomaslovato**
Based on PR #1095 create the ocean depth `mip-level` fixes for other problematic CMIP6 datastest, namely cnrm_esm2_1, gfdl_esm4, ipsl_cm6a_lr

Closes #1069 (and also #527 and #1100)

Link to documentation:

***

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [ ] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [ ] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [ ] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

**Comment by tomaslovato** [comment]
@schlunma could you please give me a feedback on `test_cnrm_esm2_1.py` before I continue with the other datasets (I feel like something is missing)... Thanks!

**Comment by schlunma** [comment]
Looks good already! The only thing that's missing is a test of the actual fix, so something like this:

https://github.com/ESMValGroup/ESMValCore/blob/4e9eb33d083a1495471a914e135c4be6bf2dbc34/tests/integration/cmor/_fixes/cmip6/test_cesm2.py#L228-L253

and this

https://github.com/ESMValGroup/ESMValCore/blob/4e9eb33d083a1495471a914e135c4be6bf2dbc34/tests/integration/cmor/_fixes/cmip6/test_cesm2.py#L310-L325

I think you can just copy-pasting these lines should work!

**Comment by tomaslovato** [comment]
@schlunma would it make this PR too long if I include also changes for MCM-Ua-1-0 dataset (#527)?

**Comment by schlunma** [comment]
No, that should be fine! Just let me know when it's ready for review :+1:

**Comment by tomaslovato** [comment]
@schlunma @remi-kazeroni the code should is ready for review 🚀 . 
I already made a test with a recipe to check that it works for ocean variables `thetao` and `no3`.
Note that for **mcm_ua_1_0** I decided to integrate needed changes in the original code under the `Allvars` fix.

**Comment by schlunma** [comment]
Great stuff! Having a look at it right now!

**Comment by remi-kazeroni** [comment]
> 
> 
> @schlunma @remi-kazeroni the code is ready for review 🚀 .
> I already made a test with a recipe to check that it works for ocean variables `thetao` and `no3`.
> Note that for **mcm_ua_1_0** I decided to integrate needed changes in the original code under the `Allvars` fix.

Thanks a lot! I'll have a look later today, in particular for the `MCM_UA_1_0` fixes.

File: tests/integration/cmor/_fixes/cmip6/test_mcm_ua_1_0.py [hunk 1]
```diff
@@ -157,3 +158,61 @@ def test_tas_fix_metadata(cubes):
...
+    fix = AllVars(None)
+    out_cubes = fix.fix_metadata(thetao_cubes)
+    assert out_cubes is thetao_cubes
+    assert len(out_cubes) == 1
+    out_cube = out_cubes[0]
+
+    # Check data of longitude
+    lon_coord = out_cube.coord('longitude')
+    assert lon_coord.points[0] == 0.9375
+    assert np.all(lon_coord.bounds[0] == [0., 1.875])
```
Comment (schlunma) [inline tests/integration/cmor/_fixes/cmip6/test_mcm_ua_1_0.py:210]: I think that test is wrong here. If you roll the data by `-1` (i.e. you shift all values to the right by 1) and do the same to the `longitude` (plus adding 360 to all negative values), you would end up with `[357.1875, 359.0625]` (starting from `[-0.9375, 357.1875]`]). Similarly, the bounds should read `[[356.25, 358.125], [358.125, 360.0]]`.

For testing float arrays, you should always use `np.testing.assert_allclose`:

```suggestion
    np.testing.assert_allclose(lon_coord.points, [357.1875, 359.0625])
    np.testing.assert_allclose(lon_coord.bounds,
                               [[356.25, 358.125], [358.125, 360.0]])
```

File: tests/integration/cmor/_fixes/cmip6/test_mcm_ua_1_0.py [hunk 2]
```diff
@@ -157,3 +158,61 @@ def test_tas_fix_metadata(cubes):
...
+
+
+def test_thetao_fix_metadata(thetao_cubes):
+    """Test ``fix_metadata`` for ``thetao``."""
+    fix = AllVars(None)
+    out_cubes = fix.fix_metadata(thetao_cubes)
+    assert out_cubes is thetao_cubes
+    assert len(out_cubes) == 1
+    out_cube = out_cubes[0]
+
```
Comment (schlunma) [inline tests/integration/cmor/_fixes/cmip6/test_mcm_ua_1_0.py:206]: You would also like to test that the data itself has been shifted correctly:

```suggestion

    # Check data of cube
    np.testing.assert_allclose(out_cube.data,
                               [[[[1, 0],
                                  [3, 2]],
                                 [[5, 4],
                                  [7, 6]]],
                                [[[9, 8],
                                  [11, 10]],
                                 [[13, 12],
                                  [15, 14]]]])
```

File: tests/integration/cmor/_fixes/cmip6/test_mcm_ua_1_0.py [hunk 3]
```diff
@@ -157,3 +158,61 @@ def test_tas_fix_metadata(cubes):
...
+        var_name='lev', standard_name=None, units='m',
+        attributes={'positive': 'up'})
+    coord_specs = [
+        (time_coord, 0),
+        (lev_coord, 1),
+        (lat_coord, 2),
+        (lon_coord, 3),
+    ]
+    thetao_cube = iris.cube.Cube(
+        np.ones((2, 2, 2, 2)),
```
Comment (schlunma) [inline tests/integration/cmor/_fixes/cmip6/test_mcm_ua_1_0.py:185]:
```suggestion
        np.arange(16).reshape(2, 2, 2, 2),
```

For testing purposes later.

**Review by schlunma — COMMENTED** [review]
Nice work @tomaslovato, looks really good already!

I have some suggestions to the fix of MCM and its test.

**Comment by tomaslovato** [comment]
@schlunma I commited all the review changes in d9988fb

**Comment by tomaslovato** [comment]
@schlunma I got a bit confused about `test_mcm_ua_1_0.py` (I'm not skilled enough on this) as it is by now it only goes through `Omon` without applying changes from `AllVars` to roll the cube and coordinates... any suggestion?

**Comment by schlunma** [comment]
No worries! There are two options here:

1. Split the test function into two tests (maybe `test_omon_thetao_fix_metadata` and `test_allvars_thetao_fix_metadata`), use `fix = Omon(...)` and `fix = AllVars(...)`, respectively, and only inlcude the `assert`s that belong to the respective fix.
2. Apply both fixes in the current test function:

```py
    fix_omon = Omon(vardef)
    fix_allvars = AllVars(vardef)
    out_cubes = fix_omon.fix_metadata(thetao_cubes)
    out_cubes = fix_allvars.fix_metadata(out_cubes)
```

Option 2 is probably easier to implement, option 1 clearly seperates the two individual fixes. Optimally we would have both tests (i.e. option 1 and 2), but I'm fine with whatever you like to implement.

**Comment by tomaslovato** [comment]
> ```python
>     fix_omon = Omon(vardef)
>     fix_allvars = AllVars(vardef)
>     out_cubes = fix_omon.fix_metadata(thetao_cubes)
>     out_cubes = fix_allvars.fix_metadata(out_cubes)
> ```

I think that for the time being option 2 will be ok, as it applies the fix to thetao including both `Omon` and `AllVars` as it should do when calling it from a recipe ...

**Comment by schlunma** [comment]
Tests are passing now, thank you really much for the fixes @tomaslovato! 🎉 

Would you mind if I push 1-2 commits to this branch directly? I just realized that a test for `Fgco2` of GFDL is missing (which is also missing for CESM2 and which I wanted to add for a long time) and that I forgot to add some documentation about the `mip`-level fixes.

I will approve afterwards and merge when @remi-kazeroni had the chance to test the fixes. Cheers!

**Comment by tomaslovato** [comment]
@schlunma thank you for the nice interaction! About you addition just go for it ! 
I've a little request ... could you have a look to #1097, that is only one line code change leftover from the former objective of this issue.

**Review by schlunma — APPROVED** [review]
@remi-kazeroni Please merge when you tested successfully :+1:

**Comment by tomaslovato** [comment]
While working on the `time` coordinate issue for climate_statistic I found a problem with the dataset fix for **ipsl_cm6a_lr**.

The `cell_area` variable was oddly added to almost all ocean variables of this dataset, so a specific fix was setup about 2 years ago
https://github.com/ESMValGroup/ESMValCore/blob/79ee767b78fd2cc25471afbc0fa7b1bc116d7543/esmvalcore/cmor/_fixes/cmip6/ipsl_cm6a_lr.py#L27-L47

After the introduction of  cube `cell_measures` trough [_ancillary_vars.py](https://github.com/ESMValGroup/ESMValCore/blob/79ee767b78fd2cc25471afbc0fa7b1bc116d7543/esmvalcore/preprocessor/_ancillary_vars.py), the ingestion of `areacello` from this dataset fails with an empty cubelist error.
Looking at the dump of the areacello file (here below), You will see that there are two variables (cell_area, areacello) both having as standard_name `cell_area` which is used by iris to assign cube name. 
The general fix control https://github.com/ESMValGroup/ESMValCore/blob/79ee767b78fd2cc25471afbc0fa7b1bc116d7543/esmvalcore/cmor/_fixes/cmip6/ipsl_cm6a_lr.py#L41-L42
 is responsible for trashing both variables and produce an empty cubelist.

For me the simplest solution is to modify the if statement as in the following
``` if cube.name() == 'cell_area' and cube.var_name != 'areacello':```

but probably we should also think to remove the creation of the auxcoord  'cell_area' (likely kept in there at the time for some potential use) which is a potentially misleading duplicate of the cube cell_measure property. 

<details>
<summary>areacello ipsl_cm6a_lr</summary>
<br>

```
netcdf areacello_Ofx_IPSL-CM6A-LR_historical_r1i1p1f1_gn {
dimensions:
	axis_nbounds = 2 ;
	x = 362 ;
	y = 332 ;
	nvertex = 4 ;
	time = UNLIMITED ; // (0 currently)
variables:
	float nav_lat(y, x) ;
		nav_lat:standard_name = "latitude" ;
		nav_lat:long_name = "Latitude" ;
		nav_lat:units = "degrees_north" ;
		nav_lat:bounds = "bounds_nav_lat" ;
	float nav_lon(y, x) ;
		nav_lon:standard_name = "longitude" ;
		nav_lon:long_name = "Longitude" ;
		nav_lon:units = "degrees_east" ;
		nav_lon:bounds = "bounds_nav_lon" ;
	float bounds_nav_lon(y, x, nvertex) ;
	float bounds_nav_lat(y, x, nvertex) ;
	float area(y, x) ;
		area:standard_name = "cell_area" ;
		area:units = "m2" ;
		area:coordinates = "nav_lon nav_lat" ;
	float areacello(y, x) ;
		areacello:standard_name = "cell_area" ;
		areacello:long_name = "Grid-Cell Area" ;
		areacello:units = "m2" ;
		areacello:online_operation = "once" ;
		areacello:cell_methods = "area: sum" ;
		areacello:cell_measures = "area: area" ;
		areacello:_FillValue = 1.e+20f ;
		areacello:missing_value = 1.e+20f ;
		areacello:coordinates = "nav_lat nav_lon" ;
		areacello:description = "Cell areas for any grid used to report ocean variables and variables which are requested as used on the model ocean grid (e.g. hfsso, which is a downward heat flux from the atmosphere interpolated onto the ocean grid). These cell areas should be defined to enable exact calculation of global integrals (e.g., of vertical fluxes of energy at the surface and top of the atmosphere)." ;
		areacello:history = "none" ;

// global attributes:
		:Conventions = "CF-1.7 CMIP-6.2" ;
		:creation_date = "2018-07-11T07:26:30Z" ;
		:tracking_id = "hdl:21.14100/718e5096-9a8f-4794-a867-0ac583371b62" ;
		:description = "CMIP6 historical" ;
		:title = "IPSL-CM6A-LR model output prepared for CMIP6 / CMIP historical" ;
		:activity_id = "CMIP" ;
...
```
</details>

**Comment by schlunma** [comment]
Good find @tomaslovato! I just pushed a tiny fix for this, please let me know if it works for you! As you said, we do not need the `cell_area` coordinate since this is included via `cell_measures`.

**Comment by tomaslovato** [comment]
@schlunma I re-run the recipe with the modified code and it works great!

**Comment by schlunma** [comment]
Great! Ready to merge from my side

**Comment by remi-kazeroni** [comment]
> 
> 
> @remi-kazeroni, could you please have a quick look?

Yes, I'm on it. I'm done opening issues and PRs for the Core. It's time to start closing them 🍺

**Review by remi-kazeroni — APPROVED** [review]
Everything looks and works fine! Thanks a lot for your work @tomaslovato 👍 This can be merged

**Pull request merged**